llama_index/vector_stores/cassandra.py

```diff
@@ -0,0 +1,248 @@
+        ttl_seconds (Optional[int]): expiration time for inserted entries.
+            Default is no expiration.
+
+    """
+
+    stores_text: bool = True
+    flat_metadata: bool = True
+
+    def __init__(
+        self,
+        session: Any,
+        keyspace: str,
+        table: str,
+        embedding_dimension: int,
+        ttl_seconds: Optional[int] = None,
+    ) -> None:
+        import_err_msg = "`cassio` package not found, please run `pip install cassio`"
+        try:
+            import cassio  # noqa: F401
+        except ImportError:
+            raise ImportError(import_err_msg)
+
+        self._session = session
+        self._keyspace = keyspace
+        self._table = table
+        self._embedding_dimension = embedding_dimension
+        self._ttl_seconds = ttl_seconds
+
+        _logger.debug("Creating VectorTable")
+        self.vector_table = cassio.vector.VectorTable(
+            session=self._session,
+            keyspace=self._keyspace,
+            table=self._table,
+            embedding_dimension=self._embedding_dimension,
+            primary_key_type="TEXT",
+        )
+
+    def add(
+        self,
+        embedding_results: List[NodeWithEmbedding],
+    ) -> List[str]:
+        """Add embedding results to index.
+
+        Args
+            embedding_results: List[NodeWithEmbedding]: list of embedding results
+
+        """
+        (document_ids0, documents, document_metadatas, embeddings,) = zip(
+            *(
+                (
+                    result.id,
+                    # The name of the key containing the text is a constant, so:
+                    getattr(result.node, DEFAULT_TEXT_KEY, ""),
+                    node_to_metadata_dict(
+                        result.node,
+                        remove_text=False,  # must keep the text here for retrieval
+                        flat_metadata=self.flat_metadata,
+                    ),
+                    result.embedding,
+                )
+                for result in embedding_results
+            )
+        )
+        document_ids = list(document_ids0)
+
+        # TODO: batching or concurrent inserts
+        _logger.debug(f"Adding {len(document_ids)} rows to table")
+        for (doc_id, doc, md, emb) in zip(
+            document_ids, documents, document_metadatas, embeddings
+        ):
+            self.vector_table.put(
+                document=doc,
+                embedding_vector=emb,
+                document_id=doc_id,
+                metadata=md,
+                ttl_seconds=self._ttl_seconds,
+            )
+
+        return document_ids
+
+    def delete(self, ref_doc_id: str, **delete_kwargs: Any) -> None:
+        """
+        Delete nodes using with ref_doc_id.
+
+        Args:
+            ref_doc_id (str): The doc_id of the document to delete.
+
+        """
+        _logger.debug("Deleting a row from VectorTable")
+        self.vector_table.delete(
+            document_id=ref_doc_id,
```

Comment:
so, when `Documents` are inserted into llama_index, they are broken/chunked into node objects. Each node has a `node.ref_doc_id` that references the ID of the document it came from.

So the delete function here is intended to work at the `ref_doc_id` level, since that is usually what users interact with the most. 

However, in this case, `document_id` seems to actually be the node id when I look at the add function, so this is not actually implemented as intended

Reply:
Acknowledged. 

@hemidactylus Brainstorming here. I think we can map a LI (LlamaIndex) Document to a Partition and a LI Node to a Row - then when we are adding new chunks as nodes, each node is co-located. 

So the CQL structure may look like 

`
CREATE TABLE ks.vs_table (
    document_id text,
    node_id text, 
    document text,
    embedding_vector vector<float, 1536>,
    metadata_blob text,
    PRIMARY KEY (document_id, node_id)
) 
`
Then we add SAI for the embedding_vector for ANN, and then SAI to the node_id for easy delete?

Reply:
Got it.
Sorry about the method's behaviour, it was a misunderstanding on my part.
We are currently looking for the optimal way to comply with this requirement while preserving commonalities with our general approach to modeling "piece-wise" storage of "documents" in a vector-capable Cassandra table.